src/remote/provisioning.rs

```diff
@@ -0,0 +1,1436 @@
+const ERR_1101: &str = "already processing request";
+const ERR_1102: &str = "version number error";
+const ERR_1103: &str = "unrecognized request type";
+const ERR_1104: &str = "request scheduled for processing";
+const ERR_1201: &str = "request - no such resource class";
+const ERR_1202: &str = "request - no resources allocated in resource class";
+const ERR_1203: &str = "request - badly formed certificate request";
+const ERR_1204: &str = "request - already used key in request";
+const ERR_1301: &str = "revoke - no such resource class";
+const ERR_1302: &str = "revoke - no such key";
+const ERR_2001: &str = "Internal Server Error - Request not performed";
+
+
+// Content-type for HTTP(s) exchanges
+pub const CONTENT_TYPE: &str = "application/rpki-updown";
+
+//------------ Message -------------------------------------------------------
+
+/// This type represents all Provisioning Messages defined in RFC 6492.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Message {
+    sender: Sender,
+    recipient: Recipient,
+    payload: Payload,
+}
+
+/// # Encoding to XML
+/// 
+impl Message {
+    /// Writes the Message's XML representation.
+    pub fn write_xml(
+        &self, writer: &mut impl io::Write
+    ) -> Result<(), io::Error> {
+        let mut writer = xml::encode::Writer::new(writer);
+
+        writer.element(MESSAGE.into())?
+            .attr("xmlns", NS)?
+            .attr("version", VERSION)?
+            .attr("sender", &self.sender)?
+            .attr("recipient", &self.recipient)?
+            .attr("type", self.payload.payload_type())?
+            .content(|content| self.payload.write_xml(content) )?;
+
+        writer.done()
+    }
+
+    /// Writes the Message's XML representation to a new String.
+    pub  fn to_xml_string(&self) -> String {
+        use std::str::from_utf8;
+
+        let mut vec = vec![];
+        self.write_xml(&mut vec).unwrap(); // safe
+        let xml = from_utf8(vec.as_slice()).unwrap(); // safe
+
+        xml.to_string()
+    }
+}
+
+/// # Decoding from XML
+/// 
+impl Message {
+    /// Parses an RFC 6492 <message />
+    pub fn decode<R: io::BufRead>(reader: R) -> Result<Self, Error> {
+        
+        let mut reader = xml::decode::Reader::new(reader);
+
+        let mut sender: Option<Sender> = None;
+        let mut recipient: Option<Recipient> = None;
+        let mut payload_type: Option<String> = None;
+
+        let mut outer = reader.start(|element| {
+            if element.name().local() != MESSAGE {
+                return Err(XmlError::Malformed)
+            }
+            
+            
+            element.attributes(|name, value| match name {
+                b"version" => {
+                    if value.ascii_into::<String>()? != VERSION {
+                        return Err(XmlError::Malformed)
+                    }
+                    Ok(())
+                }
+                b"sender" => {
+                    sender = Some(value.ascii_into()?);
```

Comment:
If an attribute appears multiple times, the later occurrences overwrite the earlier ones. That might lead to surprising errors, so maybe error out if `sender.is_some()` (and the equivalent everywhere further down)?

Reply:
Would it even be valid XML if attributes by the same name are repeated? If not then I think that `decode::attributes` should error out instead.

I can change the code here but it would become very wordy...

If we should in principle allow multiple occurrences of the same attribute, how about adding a new function like `decode::attributes_unique` or some such?